sql/plan/merge_join.go

```diff
@@ -135,13 +147,37 @@ const (
 	msSelect
 	msRet
 	msRetLeft
+	msRejectNull
 )
 
 func (i *mergeJoinIter) Next(ctx *sql.Context) (sql.Row, error) {
 	var err error
 	var ret sql.Row
 	var res int
 
+	//  The common inner join match flow:
+	//  1) check for io.EOF
+	//  2) evaluate compare filter
+	//  3) evaluate select filters
+	//  4) initialize match state
+	//  5) drain match state
+	//  6) repeat
+	//
+	// Left-join matching is unique. At any given time, we need to know whether
+	// a unique left row: 1) has already matched, 2) has more right rows
+	// available for matching before we can return a nullified-row. Otherwise
+	// we may accidentally return nullified rows that have matches (before or
+	// after the current row), or fail to return a nullified row that has no
+	// matches.
+	//
+	// We use two variables to manage the lookahead state management.
+	// |matchedleft| is a forward-looking indicator of whether the current left
+	// row has satisfied a join condition. It is reset to false when we
+	// increment left. |matchincleft| is true when the most recent call to
+	// |incmatch| incremented the left row. The two vars combined let us
+	// lookahead during msSelect to 1) identify proper nullified row matches,
+	// and 2) maintain forward-looking state for the next |i.fullrow|.
+	//
```

Comment:
Ohmygosh! These are excellent docs. 👏  Super, super helpful to very quickly understand the high level intent of this code. Very nice! 😁